cpp/include/cugraph/algorithms.hpp

```diff
@@ -24,12 +24,135 @@
+ *                                     [dim = `message_flow_graph.n_in_nodes x dimension`]
+ * @param[out] output_embedding        output aggregated embeddings.
+ *                                     [dim = `message_flow_graph.n_out_nodes x leading_dimension`]
+ * @param[out] output_extrema_location (optional) location of the max/min embeddings.
+ *                                     Valid only for kMin/kMax aggregations
+ *                                     [dim = `message_flow_graph.n_out_nodes x dimension`]
+ */
+
+template <typename vertex_t>
+void aggregate_concatenate_forward(raft::handle_t const& handle,
+                                   ops::gnn::graph::mfg_csr<vertex_t> const& message_flow_graph,
+                                   size_t dimension,
+                                   ops::gnn::AggOpT op,
+                                   float const* input_embedding,
+                                   float* output_embedding,
+                                   std::optional<vertex_t*> output_extrema_location);
+
+/**
+ * @brief Computes the backward pass for aggregation layer
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ *
+ * @param handle                     RAFT handle object to encapsulate resources (e.g. CUDA stream,
+ *                                   communicator, and handles to various CUDA libraries) to run
+ * graph algorithms.
+ * @param[in] message_flow_graph     input message flow graph.
+ * @param[in] dimension              dimensionality of the embeddings
+ * @param[in] leading_dimension      leading dimension for `input_gradient` and
+ *                                   `output_gradient` array. It must be at least `dimension`.
+ * @param[in] ignore_self            ignore the self node in the aggregation
+ * @param[in] op                     aggregation operation
+ * @param[in] input_gradient         output aggregated embeddings.
+ *                                   [dim = `message_flow_graph.n_out_nodes x leading_dimension`]
+ * @param[in] input_extrema_location (optional) location of the max/min embeddings. This will
+ *                                   be written to only for kMin/kMax aggregations
+ *                                   [dim = `message_flow_graph.n_out_nodes x dimension`]
+ * @param[out] output_gradient       input embeddings.
+ *                                   [dim = `message_flow_graph.n_in_nodes x dimension`]
+ */
+template <typename vertex_t>
+void aggregate_backward(raft::handle_t const& handle,
+                        ops::gnn::graph::mfg_csr<vertex_t> const& message_flow_graph,
+                        size_t dimension,
+                        size_t leading_dimension,
+                        bool ignore_self,
+                        ops::gnn::AggOpT op,
+                        float const* input_gradient,
+                        std::optional<vertex_t const*> input_extrema_location,
+                        float* output_gradient);
```

Comment:
I personally prefer renaming variables names to `input_gradients`, `output_gradients` and `input_extrema_locations` to better imply that this is for multiple values (not a pointer to a single value).